### ISSUE TITLE
Investigate Full Calendar accessibility compliance issues

### ISSUE THREAD

**Issue by briand44** [uiowa/uiowa801#9365]
# Motivation

SiteImprove is reporting a wcag 2.1 A issue with calendar displays.

>This page has a scrollable section that can't be operated by keyboard.

**Examples**

Admissions
https://my2.siteimprove.com/Inspector/998093/A11y/Page?pageId=71788717907&impmd=264ED37B65FA6186129334A508391970&conf=a+aa#/sia-r84/failed/bVNFOUdSMDc=;MTM2ZGVkNzg=

Grad
https://my2.siteimprove.com/Inspector/998426/A11y/Page?pageId=29973646729&impmd=264ED37B65FA6186129334A508391970&conf=a+aa#/sia-r84/failed/ZkQ5NTFybFk=;NDg3YjE5NzI=

Related issue #8497

ITS
https://my2.siteimprove.com/Inspector/994335/A11y/Page?pageId=28396872889&impmd=264ED37B65FA6186129334A508391970&conf=a#/sia-r84/failed

---

Also seeing another issue reported related to calendar displays: No data cells assigned to table header

>Every header cell in a table should have at least one data cell assigned to it.

https://my2.siteimprove.com/Inspector/998093/A11Y/Page?pageId=71788717907&impmd=264ED37B65FA6186129334A508391970&issue=failed&decision=dismissed+falsePositive+none&comp=ada#/sia-r46/failed/TGFBWVhmbzM=;ZDVhNmI2ZDY=

**Comment by briand44** [comment]
Scheduled time to meet and discuss with ITS. Reached out to Grad College to see if any updates on the MAUI side of things.

**Comment by briand44** [comment]
Met with Grad College... They are still not sure when MAUI will be ready to pull data directly (https://github.com/uiowa/uiowa/issues/8497) but will try to get more clarity there. Discussed replacing the calendar display with a searchable/filterable display of data that would display the data from the website database until MAUI is ready. Once MAUI ready, could use the same display just either import from MAUI first or pull directly from MAUI. Seemed happy with this approach but plan to get feedback from other stakeholders.

**Comment by briand44** [comment]
ITS discussed with their leadership and they really like the calendar display and would prefer to keep if possible. Sounds like they had some conversations with someone from the accessibility task force and have the understanding that if the list view is accessible they would be considered compliant because there is an accessible alternative. They would be open to defaulting to the list view rather than the monthly calendar display if needed to be compliant.

**Comment by briand44** [comment]
Discussed this today and have a few takeaways:
- Compliance first, alternate versions if compliance not possible
  - We should invest in updating or moving to module to see if we can achieve compliance since ITS prefers to keep calendar display
- Full calendar list display not great in terms of accessibility either
  - Data is output in a table - adding `role="presentation"` made the screen reader experience somewhat better
  - keyboard experience getting to list not great
  - Can we give users option of list display vs month (calendar) earlier

**Comment by briand44** [comment]
## Grooming
 
**Quadrant:** 

Complex

 **Discussion:**
 - previously flipped from one module to another
 - may need to evaluate these modules plus library directly
 - full calendar module vs full calendar views - One is now in the events recipe within Drupal CMS
 - switching modules may get us closer but may not get us where we need to be
 - if we need to rely on alternate, could disable full calendar list and link to accessible list
   - Create grid display for https://registrar.uiowa.edu/academic-calendar as pseudo calendar?
 - does the most recent version of the library even resolve the accessibility issues?
   - if not, updating/switching modules may not be worth the effort
   - if so, investigating module updates, switching modules
     - fullcalendar_view 5.1.13 is not D11 compatible
     - fullcalendar_view has premium version that supports newer version of library
 - https://www.drupal.org/project/fullcalendar famous drupalers are hanging out here
 - may not be able to get to the latest version of library through modules
 - Siteimprove is currently reporting No data cells assigned to table header and scrollable element not keyboard accessible

**Proposed solution or definition of done:**
- evaluate latest fullcalendar library in terms of accessibility
- investigate https://www.drupal.org/project/fullcalendar
- investigate https://www.drupal.org/project/fullcalendar_views
- Provide recommendation on approach to meet accessibility compliance based on investigations and evaluations. If we don't think we can make it compliant, what would be the recommended approach for an accessible alternative?

**Comment by joewhitsitt** [comment]
Just ran the siteimprove browser extension against the demo (https://fullcalendar.io/demos). I gutted the top bar, sidebar and footer to isolate just the calendar itself. I believe it is running 6.1.20.

- Table header cell is missing a header role - A -  1.3.1 Info and Relationships (days of the week)
- ARIA attribute unsupported or prohibited - ARIA - WAI-ARIA authoring practices (days of the week links, +2 more link, etc.)
- Image missing a text alternative - A - 1.1.1 Non-text Content (forward/back arrow nav)
- Color contrast does not meet minimum requirement - AA - 1.4.3 Contrast (Minimum), AAA - 1.4.6 Contrast (Enhanced) (white on blue pills representing events).
- Text is clipped when resized - AA - 1.4.4 Resize Text (days of the week)

Switching to the list display, I only get the ARIA attribute unsupported or prohibited and Image missing a text alternative issues.

So TLDR; newest stable version of fullcalendar does not solve all of our problems. Tomorrow I will see if the implementation in DrupalCMS solves any of these or introduces more.

**Comment by joewhitsitt** [comment]
Dealing with local issues, I finally got the https://www.drupal.org/project/events_calendar/releases/1.0.1 recipe installed on D11.3. It uses https://www.drupal.org/project/fullcalendar

It adds primary tabs to the interface resulting in the additional accessibility errors below.

- Text not included in an ARIA landmark - ARIA - WAI-ARIA authoring practices
- Container element is empty - A - 1.3.1 Info and Relationships

<img width="570" height="264" alt="Image" src="https://github.com/user-attachments/assets/4f680669-72db-4438-b18b-3cf9c76e16ed" />

It does not feature the built-in fullcalendar list view and it does fix the color contrast issue with the event pills and the text is clipped issue with the days of the week.

<img width="270" height="165" alt="Image" src="https://github.com/user-attachments/assets/dff54b9a-f65c-4ca1-8e02-2a176480635a" />

<img width="985" height="102" alt="Image" src="https://github.com/user-attachments/assets/cf7f205d-c6f5-4599-b345-74a9271c5b8d" />

So it seems like some of the issues can be resolved from theming and not tied to the library itself

**Comment by joewhitsitt** [comment]
I've been having troubles just getting fullcalendar_view to render without a 500 error locally on its.uiowa.ddev.site. Disabling the view display block allows the page to resolve.

Unless I can figure out what is going on, I might switch gears a bit to see what a transition to fullcalendar module would look like.

UPDATE: the current view is displaying all alerts... 15 pages of them. adding a filter of start greater or equal to -90 day seems to help. might need to determine what this is and limit going backwards on the calendar.

**Comment by joewhitsitt** [comment]
tried switching out fullcalendar_view in favor of fullcalendar module. it involves updating the split and updating views config to match what was there before, but it is fairly straightforward.

noticed several of the issues seem to be remedied. one new one though:

- Scrollable element is not keyboard accessible - A - 2.1.1 Keyboard

We also need to adjust our past/future month dates to not be so light in color.

Also, it is rendering a exposed date filter but I can't figure out why.

<img width="1000" height="863" alt="Image" src="https://github.com/user-attachments/assets/63fa9c20-6e32-44a3-8e1d-b7c37bfc369f" />

**Comment by briand44** [comment]
> one new one though:
> 
> Scrollable element is not keyboard accessible - A - 2.1.1 Keyboard

Just noting that "Scrollable element is not keyboard accesible" is currently being reported in prod.

https://my2.siteimprove.com/Inspector/994335/A11Y/Page?pageId=28396872889&impmd=264ED37B65FA6186129334A508391970&conf=a+aa+aria+si&issue=failed&decision=dismissed+falsePositive+none&comp=ada#/sia-r84/failed/Tlg3MEEyZG0=;MmExODZhZjc=

**Comment by joewhitsitt** [comment]
Yeah, that seems to be a SiteNow-ism, I don't recall seeing that in the non-SiteNow implementations

**Comment by joewhitsitt** [comment]
Ajax was causing the date range exposed filter. 🤷 https://www.drupal.org/project/fullcalendar/issues/3507568

Alright. I am in the boat of "if we need to keep using full calendar, we should switch the module and provide an alternative views list display"

#9785 is the start of that work.

**Comment by joewhitsitt** [comment]
Forgot to move this to blocked waiting this morning:

- evaluate latest fullcalendar library in terms of accessibility
  - https://github.com/uiowa/uiowa/issues/9365#issuecomment-4300166625
- investigate https://www.drupal.org/project/fullcalendar
  -  https://github.com/uiowa/uiowa/issues/9365#issuecomment-4328910060
- investigate https://www.drupal.org/project/fullcalendar_views
  - a lot of the same issues plus the keyboard accessibility issue.
- Provide recommendation on approach to meet accessibility compliance based on investigations and evaluations. If we don't think we can make it compliant, what would be the recommended approach for an accessible alternative?
  - https://github.com/uiowa/uiowa/issues/9365#issuecomment-4337215817

**Comment by joewhitsitt** [comment]
Maybe looking to @briand44 on whether this is done. I might still chip away at the POC but that wasn't expected for this issue.

**Comment by joewhitsitt** [comment]
mandclu just announced dev branches for fullcalendar/fullcalendar_io that use fullcalendar v7 (not ready for production) but "...already there are a number of new theme options, some accessibility improvements, etc. We also plan to add other newly available options like dark mode support, sticky day headers, and more"

i think it is our sticky header implementation on its alerts calendar that might be causing the keyboard accessibility issue.

**Comment by pyrello** [comment]
> mandclu just announced dev branches for fullcalendar/fullcalendar_io that use fullcalendar v7 (not ready for production) but "...already there are a number of new theme options, some accessibility improvements, etc. We also plan to add other newly available options like dark mode support, sticky day headers, and more"
> 
> i think it is our sticky header implementation on its alerts calendar that might be causing the keyboard accessibility issue.

Maybe we can wait to see how this develops since we have ~11 months now to meet compliance requirements?

**Comment by briand44** [comment]
Yeah, I think we can close this. I will get a follow up created based off of the recommendations but will probably hold on prioritizing to see if/how work with v7 progresses.
